### PR TITLE
[tap] graceful failure when streaming output with static config

### PR DESCRIPTION
**Pull request by asraa**
Envoy crashes on RELEASE_ASSERT if a user specifies an streaming admin output sink with a static config. This catches the configuration failure early.

Testing: config test and regression fuzz test.

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by asraa** [comment]
QQ -- This RELEASE_ASSERT won't hit anymore in the tested codepath https://github.com/envoyproxy/envoy/blob/4b6e81c112e204ef04445318d2cbfa168cf22ca7/source/extensions/common/tap/tap_config_base.cc#L60
and ideally I'd like to get rid of it. I want to guarantee static config is the only case where we pass a nullptr admin, but we seem to lose the config type at that stage. https://github.com/envoyproxy/envoy/blob/4b6e81c112e204ef04445318d2cbfa168cf22ca7/source/extensions/common/tap/extension_config_base.cc#L31

File: test/extensions/filters/http/tap/tap_filter_test.cc [hunk 1]
```diff
@@ -126,6 +128,31 @@ TEST_F(TapFilterTest, Config) {
...
+  envoy::extensions::filters::http::tap::v3::Tap config;
+  TestUtility::loadFromYaml(filter_config, config);
+  NiceMock<Server::Configuration::MockFactoryContext> context;
+  TapFilterFactory factory;
+  EXPECT_THROW_WITH_MESSAGE(factory.createFilterFactoryFromProto(config, "stats", context),
+                            EnvoyException,
+                            "Error: Specifying admin streaming output without configuring admin.");
+  // Http::MockFilterChainFactoryCallbacks filter_callbacks;
+  // EXPECT_CALL(filter_callbacks, addStreamDecoderFilter(_));
+  // cb(filter_callbacks);
```
Comment (mattklein123) [inline test/extensions/filters/http/tap/tap_filter_test.cc:153]: ?

**Comment by mattklein123** [comment]
looks like some legit clang-tidy issues, some pre-existing. Do you mind seeing if any of them are easy to fix?

/wait

**Comment by asraa** [comment]
Will do RE clang_tidy -- it looks like TAP extensions call virtual methods in constructor, I'll resolve that, just brainstorming the most effective way

**Comment by asraa** [comment]
Will open an issue about this unless it's clear what's going on:
`newTapConfig` is a virtual method of `ExtensionConfig`, and implemented in `ExtensionConfigBase` (which has a subclass `TransportSocketFactory`.
* It's called in the constructor of an `ExtensionConfigBase` when initializing a static config
* It's also called for each registered `ExtensionConfig`  in the static `AdminHandler` handler

Is it meant to be called in the constructor at all? If so, I can move it out of the constructor call and call it in `createTransportSocketFactory` (the only place where the constructor ends up being called), but it's a little odd if I should only do it in the static config case.
https://github.com/envoyproxy/envoy/blob/fca0f0bf5e865de951f72ae839ed4487cd4bd55d/source/extensions/common/tap/extension_config_base.cc#L31

**Comment by mattklein123** [comment]
@asraa `newTapConfig` is also called in the admin handler when a new config is loaded via REST call (it would also be called in the future for other update types such as xDS update, etc.). I think the easy short-term fix is probably just having a non-virtual variant of `newTapConfig` in that class which is called from both the virtual override call as well as the extension base constructor. (The virtual call is there because the other code interacts with it via an interface.)

**Comment by mattklein123** [comment]
LGTM modulo format. Thank you!

/wait

**Pull request merged**